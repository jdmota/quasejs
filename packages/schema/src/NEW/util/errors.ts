import type { Path } from "./path";

export class SchemaError {
  readonly value: unknown;
  readonly path: Path;
  readonly message: string;
  constructor(value: unknown, path: Path, message: string) {
    this.value = value;
    this.path = path;
    this.message = message;
  }
}
