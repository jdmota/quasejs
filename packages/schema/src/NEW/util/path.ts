export class Path {
  public readonly parent: Path | null;
  public readonly key: string | number | null;
  public readonly context: string | null;

  constructor(
    parent: Path,
    key: string | number | null,
    context: string | null
  );
  constructor(parent: null, key: null, context: null);
  constructor(
    parent: Path | null,
    key: string | number | null,
    context: string | null
  ) {
    this.parent = parent;
    this.key = key;
    this.context = context;
  }

  isRoot() {
    return this.parent == null;
  }

  push(key: string | number | null, context: string | null = null) {
    return new Path(this, key, context);
  }

  pop(): Path {
    if (this.parent == null)
      throw new Error("Invariant: validation path - cannot pop the root");
    return this.parent;
  }

  static create() {
    return new Path(null, null, null);
  }
}
