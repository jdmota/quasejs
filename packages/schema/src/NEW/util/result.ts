import type { SchemaError } from "./errors";

export type ValidationOK<T> = {
  readonly ok: true;
  readonly value: T;
};

export type ValidationError = {
  readonly ok: false;
  readonly errors: readonly SchemaError[];
};

export type ValidationResult<T> = ValidationOK<T> | ValidationError;

export type ValidationResultMaybeAsync<T> =
  | ValidationResult<T>
  | Promise<ValidationResult<T>>;

export const ValidationResult = {
  ok<T>(value: T): ValidationOK<T> {
    return { ok: true, value };
  },
  error(error: SchemaError): ValidationError {
    return { ok: false, errors: [error] };
  },
  errors(errors: readonly SchemaError[]): ValidationError {
    return { ok: false, errors };
  },
};

export type TransformResult<T> =
  | {
      readonly ok: true;
      readonly value: T;
    }
  | {
      readonly ok: false;
      readonly errors: readonly SchemaError[];
    };

export const TransformResult = {
  ok<T>(value: T): TransformResult<T> {
    return { ok: true, value };
  },
  error<T>(error: SchemaError): TransformResult<T> {
    return { ok: false, errors: [error] };
  },
  errors<T>(errors: readonly SchemaError[]): TransformResult<T> {
    return { ok: false, errors };
  },
};

export function then<A, B>(
  value: A | Promise<A>,
  fn: (value: A) => B | Promise<B>
) {
  if (value instanceof Promise) {
    return value.then(fn);
  }
  return fn(value);
}
