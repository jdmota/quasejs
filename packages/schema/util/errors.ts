import type { Path } from "./path";

export class SchemaError {
  readonly path: Path;
  readonly message: string;
  constructor(path: Path, message: string) {
    this.path = path;
    this.message = message;
  }
}
