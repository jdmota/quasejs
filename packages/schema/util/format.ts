export type Formatter = (value: unknown) => string;

export function format(value: unknown) {
  if (value === null) return "null";
  if (value === undefined) return "undefined";
  if (value === "") return "an empty string";
  if (Number.isNaN(value)) return "NaN";
  return JSON.stringify(value);
}

export function formatKey(key: string) {
  return JSON.stringify(key);
}
