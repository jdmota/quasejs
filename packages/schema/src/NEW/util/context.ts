import { type JsType } from "../types/js-types";
import { SchemaError } from "./errors";
import { format, Formatter } from "./format";
import { Path } from "./path";
import { ValidationError, ValidationResult } from "./result";

export type SchemaOpCtxOpts = {
  formatter?: Formatter;
  abortEarly?: boolean;
  allowCircular?: boolean;
};

export class SchemaOpCtx {
  private path: Path;
  private readonly errorArr: SchemaError[];
  public readonly formatter: Formatter;
  public readonly abortEarly: boolean;
  public readonly allowCircular: boolean;
  private readonly busy: WeakMap<WeakKey, Set<JsType<any>>>;

  constructor(opts: SchemaOpCtxOpts | SchemaOpCtx = {}) {
    this.path = Path.create();
    this.errorArr = [];
    this.formatter = opts.formatter ?? format;
    this.abortEarly = opts.abortEarly ?? false;
    this.allowCircular = opts.allowCircular ?? false;
    this.busy = new WeakMap();
  }

  static new(ctx: SchemaOpCtxOpts | SchemaOpCtx) {
    return new SchemaOpCtx(ctx);
  }

  shouldAbort() {
    return this.abortEarly && this.errorArr.length > 0;
  }

  createError(message: string) {
    return new SchemaError(this.path, message);
  }

  addError(message: string) {
    this.errorArr.push(this.createError(message));
  }

  format(value: unknown) {
    const { formatter } = this;
    return formatter(value);
  }

  push(key: string | number, inKey = false) {
    this.path = this.path.push(key, inKey);
  }

  pop() {
    this.path = this.path.pop();
  }

  hasErrors() {
    return this.errorArr.length === 0;
  }

  getErrors(): readonly SchemaError[] {
    return this.errorArr;
  }

  error(message: string): ValidationError {
    this.addError(message);
    return ValidationResult.errors(this.errorArr);
  }

  result<T>(value: T): ValidationResult<T> {
    return this.errorArr.length === 0
      ? ValidationResult.ok(value as T)
      : ValidationResult.errors(this.errorArr);
  }

  returnErrors() {
    return ValidationResult.errors(this.errorArr);
  }

  validate<T>(
    valid: boolean,
    whatDesc: string,
    value: unknown
  ): ValidationResult<T> {
    if (!valid) {
      this.addError(`Expected ${whatDesc} (got ${this.format(value)})`);
    }
    return this.result(value as T);
  }

  transferTo(ctx: SchemaOpCtx) {
    for (const error of ctx.errorArr) {
      ctx.errorArr.push(error);
    }
  }

  resetErrors() {
    this.errorArr.length = 0;
  }

  pushValue(value: unknown, type: JsType<any>) {
    if (typeof value === "object" && value != null) {
      const seenTypes = this.busy.get(value);
      if (seenTypes) {
        if (seenTypes.has(type)) {
          if (!this.allowCircular) {
            this.addError("Circular reference");
          }
          return false;
        } else {
          seenTypes.add(type);
        }
      } else {
        this.busy.set(value, new Set([type]));
      }
    }
    return true;
  }

  popValue(value: unknown) {
    if (typeof value === "object" && value != null) {
      this.busy.delete(value);
    }
  }
}
