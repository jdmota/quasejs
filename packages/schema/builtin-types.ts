import { VALID_JS_ID } from "../util/js-identifiers";
import { UniqueNames } from "../util/unique-names";
import { SchemaType } from "./schema-type";

abstract class BuiltinSchemaType extends SchemaType {}

export type { BuiltinSchemaType };

export class UnknownType extends BuiltinSchemaType {
  static build = new UnknownType();

  readonly _tag = "UnknownType";

  override getName() {
    return "unknown";
  }
}

export class UndefinedType extends BuiltinSchemaType {
  static build = new UndefinedType();

  readonly _tag = "UndefinedType";

  override getName() {
    return "undefined";
  }
}

export class NullType extends BuiltinSchemaType {
  static build = new NullType();

  readonly _tag = "NullType";

  override getName() {
    return "null";
  }
}

export class LiteralType extends BuiltinSchemaType {
  static build(value: number | bigint | string | boolean | symbol) {
    return new LiteralType(value);
  }

  readonly _tag = "LiteralType";

  constructor(readonly value: number | bigint | string | boolean | symbol) {
    super();
    if (typeof value === "symbol") {
      if (!value.description || value !== Symbol.for(value.description)) {
        throw new Error(`Symbol should be created with Symbol.for()`);
      }
    }
  }

  override getName() {
    return "literal";
  }
}

// TODO https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html
// TODO https://www.typescriptlang.org/docs/handbook/2/indexed-access-types.html

export class StringType extends BuiltinSchemaType {
  static build = new StringType();

  readonly _tag = "StringType";

  override getName() {
    return "string";
  }
}

export class NumberType extends BuiltinSchemaType {
  static build = new NumberType();

  readonly _tag = "NumberType";

  override getName() {
    return "number";
  }
}

export class BigintType extends BuiltinSchemaType {
  static build = new BigintType();

  readonly _tag = "BigintType";

  override getName() {
    return "bigint";
  }
}

export class BooleanType extends BuiltinSchemaType {
  static build = new BooleanType();

  readonly _tag = "BooleanType";

  override getName() {
    return "boolean";
  }
}

export class SymbolType extends BuiltinSchemaType {
  static build = new SymbolType();

  readonly _tag = "SymbolType";

  override getName() {
    return "symbol";
  }
}

export class ArrayType extends BuiltinSchemaType {
  static build(element: SchemaType, readonly: boolean = true) {
    return new ArrayType(element, readonly);
  }

  readonly _tag = "ArrayType";

  constructor(
    readonly element: SchemaType,
    readonly readonly: boolean
  ) {
    super();
  }

  override getName() {
    return "array";
  }
}

type TupleItemOpt =
  | SchemaType
  | Readonly<{
      name?: string;
      type: SchemaType;
      rest?: boolean;
    }>;

type TupleItem = Readonly<{
  name: string;
  type: SchemaType;
  rest: boolean;
}>;

export class TupleType extends BuiltinSchemaType {
  static build(elements: readonly TupleItemOpt[], readonly: boolean = true) {
    return new TupleType(elements, readonly);
  }

  readonly _tag = "TupleType";
  readonly readonly: boolean;
  readonly hasRest: boolean;
  readonly elements: readonly TupleItem[];

  constructor(elements: readonly TupleItemOpt[], readonly: boolean) {
    super();
    this.readonly = readonly;
    const uniqNames = new UniqueNames();
    let hasRest = false;
    for (const item of elements) {
      if (hasRest) {
        throw new Error(`Only 1 rest argument allowed at the end`);
      }
      if (!(item instanceof SchemaType)) {
        if (item.name) {
          if (!uniqNames.mark(item.name)) {
            throw new Error(`Duplicate tuple item name: ${item.name}`);
          }
          if (!VALID_JS_ID.test(item.name)) {
            throw new Error(`Invalid tuple item name: ${item.name}`);
          }
        }
        if (item.rest) {
          hasRest = true;
        }
      }
    }
    this.hasRest = hasRest;
    this.elements = elements.map((item, i) => {
      if (item instanceof SchemaType) {
        return {
          name: uniqNames.newInternal(`_arg${i}`),
          type: item,
          rest: false,
        };
      }
      return {
        name: item.name ?? uniqNames.newInternal(`_arg${i}`),
        type: item.type,
        rest: item.rest ?? false,
      };
    });
  }

  override getName() {
    return "tuple";
  }
}

type ObjStructure = {
  readonly [key: string]:
    | SchemaType
    | Readonly<{
        type: SchemaType;
        readonly?: boolean;
        partial?: boolean;
      }>;
};

type ObjEntries = readonly (readonly [
  string,
  Readonly<{
    readonly: boolean;
    partial: boolean;
    type: SchemaType;
  }>,
])[];

type UnknownKeysOpts = Readonly<{
  key: SchemaType;
  value: SchemaType;
  readonly?: boolean;
  partial?: boolean;
}>;

type UnknownKeys = Readonly<{
  key: SchemaType;
  value: SchemaType;
  readonly: boolean;
  partial: boolean;
}>;

const hasOwn = Object.prototype.hasOwnProperty;
const hasProp = (o: any, k: string) => hasOwn.call(o, k);
const getProp = (o: any, k: string) => (hasProp(o, k) ? o[k] : undefined);

const PROTO_KEY = "__proto__";

export class ObjectType extends BuiltinSchemaType {
  static build(
    structure: ObjStructure,
    exact: boolean | UnknownKeysOpts = true
  ) {
    return new ObjectType(structure, exact);
  }

  readonly _tag = "ObjectType";
  readonly entries: ObjEntries;
  readonly exact: boolean | UnknownKeys;

  constructor(structure: ObjStructure, exact: boolean | UnknownKeysOpts) {
    super();
    if (hasProp(structure, PROTO_KEY)) {
      throw new Error("Object type includes __proto__ key");
    }
    this.entries = Object.entries(structure).map(([k, v]) => {
      if (v instanceof SchemaType) {
        return [k, { type: v, readonly: true, partial: false }];
      }
      return [
        k,
        {
          type: v.type,
          readonly: v.readonly ?? true,
          partial: v.partial ?? false,
        },
      ];
    });
    this.exact =
      typeof exact === "boolean"
        ? exact
        : {
            key: exact.key,
            value: exact.value,
            readonly: exact.readonly ?? true,
            partial: exact.partial ?? false,
          };
  }

  override getName() {
    return "object";
  }
}

export class RecordType extends BuiltinSchemaType {
  static build(key: SchemaType, value: SchemaType, readonly = true) {
    return new RecordType(key, value, readonly);
  }

  readonly _tag = "RecordType";

  constructor(
    readonly key: SchemaType,
    readonly value: SchemaType,
    readonly readonly: boolean
  ) {
    super();
    // TODO validate kind of key: string | number | symbol
  }

  override getName() {
    return "record";
  }
}

export class UnionType extends BuiltinSchemaType {
  static build(items: readonly SchemaType[]) {
    return new UnionType(items);
  }

  readonly _tag = "UnionType";

  constructor(readonly items: readonly SchemaType[]) {
    super();
  }

  override getName() {
    return "union";
  }
}

export class IntersectionType extends BuiltinSchemaType {
  static build(items: readonly SchemaType[]) {
    return new IntersectionType(items);
  }

  readonly _tag = "IntersectionType";

  constructor(readonly items: readonly SchemaType[]) {
    super();
  }

  override getName() {
    return "intersection";
  }
}

export class FunctionType extends BuiltinSchemaType {
  static build(args: readonly TupleItemOpt[], ret: SchemaType) {
    return new FunctionType(args, ret);
  }

  readonly _tag = "FunctionType";
  readonly args: TupleType;
  readonly ret: SchemaType;

  constructor(args: readonly TupleItemOpt[], ret: SchemaType) {
    super();
    this.ret = ret;
    this.args = TupleType.build(args, true);
  }

  override getName() {
    return "function";
  }
}

type EnumLike = EnumObj | readonly string[];

type EnumObj = {
  readonly [k: string]: string | number;
};

export class EnumType extends BuiltinSchemaType {
  static build(enumeration: EnumLike) {
    return new EnumType(enumeration);
  }

  readonly _tag = "EnumType";
  readonly obj: EnumObj;
  readonly values: readonly (string | number)[];

  constructor(enumeration: EnumLike) {
    super();
    this.obj = Array.isArray(enumeration)
      ? Object.fromEntries(enumeration.map((v, i) => [v, i]))
      : enumeration;
    this.values = Object.values(this.obj);
  }

  override getName() {
    return "enum";
  }
}

export class RecursiveType extends BuiltinSchemaType {
  static build(fn: (that: RecursiveType) => SchemaType) {
    return new RecursiveType(fn);
  }

  readonly _tag = "RecursiveType";
  public readonly content: SchemaType;

  constructor(readonly fn: (that: RecursiveType) => SchemaType) {
    super();
    // TODO check guardedness
    this.content = fn(this);
  }

  override getName() {
    return "recursive";
  }
}

// TODO generics

export const builtin = {
  unknown: UnknownType.build,
  undefined: UndefinedType.build,
  null: NullType.build,
  literal: LiteralType.build,
  string: StringType.build,
  number: NumberType.build,
  bigint: BigintType.build,
  symbol: SymbolType.build,
  array: ArrayType.build,
  tuple: TupleType.build,
  object: ObjectType.build,
  record: RecordType.build,
  union: UnionType.build,
  inter: IntersectionType.build,
  func: FunctionType.build,
  enum: EnumType.build,
  rec: RecursiveType.build,
} as const;

export type TypesBuilder = typeof builtin;

export type BuiltinTypesMap = {
  [key in keyof TypesBuilder]: TypesBuilder[key] extends (...args: any[]) => any
    ? ReturnType<TypesBuilder[key]>
    : TypesBuilder[key];
};

export type BuiltinTypes = BuiltinTypesMap[keyof BuiltinTypesMap];

export function isBuiltinType(schema: SchemaType): schema is BuiltinTypes {
  return schema instanceof BuiltinSchemaType;
}
