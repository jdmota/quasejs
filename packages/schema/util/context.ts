import { assertion } from "../../util/miscellaneous";
import { SchemaError } from "./errors";
import { format, type Formatter } from "./format";
import { Path } from "./path";
import { type ValidationError, ValidationResult } from "./result";

export type SchemaOpCtxOpts = {
  readonly formatter?: Formatter;
  readonly abortEarly?: boolean;
};

export class SchemaOpCtx implements SchemaOpCtxOpts {
  private path: Path;
  private readonly errorArr: SchemaError[];
  public readonly formatter: Formatter;
  public readonly abortEarly: boolean;
  private readonly busy: WeakSet<WeakKey>;

  constructor(opts: SchemaOpCtxOpts | SchemaOpCtx = {}) {
    this.path = Path.create();
    this.errorArr = [];
    this.formatter = opts.formatter ?? format;
    this.abortEarly = opts.abortEarly ?? true;
    this.busy = new WeakSet();
  }

  static new(ctx: SchemaOpCtxOpts | SchemaOpCtx = {}) {
    return new SchemaOpCtx(ctx);
  }

  shouldAbort() {
    return this.abortEarly && this.errorArr.length > 0;
  }

  createError(message: string) {
    return new SchemaError(this.path, message);
  }

  addError(message: string) {
    this.errorArr.push(this.createError(message));
  }

  format(value: unknown) {
    const { formatter } = this;
    return formatter(value);
  }

  push(key: string | number | null, context: string | null = null) {
    this.path = this.path.push(key, context);
  }

  pop() {
    this.path = this.path.pop();
  }

  isOK() {
    return this.errorArr.length === 0;
  }

  hasErrors() {
    return this.errorArr.length > 0;
  }

  error(message: string): ValidationError {
    this.addError(message);
    return ValidationResult.errors(this.errorArr);
  }

  returnErrors() {
    assertion(this.hasErrors());
    return ValidationResult.errors(this.errorArr);
  }

  result<T>(value: T): ValidationResult<T> {
    return this.errorArr.length === 0
      ? ValidationResult.ok(value as T)
      : ValidationResult.errors(this.errorArr);
  }

  resetErrors() {
    this.errorArr.length = 0;
  }

  pushValue(value: unknown) {
    if (typeof value === "object" && value != null) {
      if (this.busy.has(value)) {
        return false;
      }
      this.busy.add(value);
    }
    return true;
  }

  popValue(value: unknown) {
    if (typeof value === "object" && value != null) {
      this.busy.delete(value);
    }
  }
}
