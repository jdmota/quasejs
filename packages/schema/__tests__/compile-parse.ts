import { beforeAll, expect, it } from "@jest/globals";
import { builtin as t } from "../builtin-types";
import {
  compileParse,
  registerParseCompilers,
} from "../compilers/compile-parse";
import { SchemaOpCtx } from "../util/context";

beforeAll(async () => {
  await registerParseCompilers();
});

it("compile parse example", () => {
  const obj = t.object({
    a: t.null,
    b: t.number,
    c: t.array(t.string),
    d: t.object(
      {},
      {
        key: t.string,
        value: t.bigint,
      }
    ),
  });

  const compiled = compileParse(obj);

  expect(compiled).toMatchSnapshot();

  const parse = compiled.makeFunc();

  expect(
    parse(
      {
        a: undefined,
      },
      SchemaOpCtx.new()
    )
  ).toMatchSnapshot();
});
