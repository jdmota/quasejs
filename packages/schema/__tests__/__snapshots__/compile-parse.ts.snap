// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`compile parse example 1`] = `
{
  "contents": "const parseCircular = (fn, value, ctx) => {
  if (ctx.pushValue(value)) {
    const r = fn(value, ctx);
    ctx.popValue(value);
    return r;
  }
  return ctx.error("Circular reference disallowed");
};
const hasOwn = Object.prototype.hasOwnProperty;
const hasProp = (o, k) => hasOwn.call(o, k);
const PROTO_KEY = "__proto__";
const getProp = (o, k) => (hasProp(o, k) ? o[k] : undefined);
const catchUnknownKeys = (object, ctx, newEntries, extraneousKeys, keyParse, valueParse, partial) => {
  for (const key of extraneousKeys) {
    ctx.push(key, "key");
    const keyResult = keyParse(key, ctx);
    ctx.pop();
    if (ctx.shouldAbort()) return ctx.returnErrors();
    ctx.push(key, "value");
    const value = getProp(object, key);
    if (!partial || value !== undefined) {
      const valueResult = valueParse(value, ctx);
      if (keyResult.ok && valueResult.ok) {
        newEntries.push([keyResult.value, valueResult.value]);
      }
    }
    ctx.pop();
    if (ctx.shouldAbort()) return ctx.returnErrors();
  }
};
const formatKey = key => JSON.stringify(key);
const reportExtraneousKeys = (ctx, extraneousKeys) => {
  return ctx.error(
    \`Extraneous properties: \${Array.from(extraneousKeys)
      .map(k => formatKey(k))
      .join(", ")}\`
  );
};
const parse_object = (value, ctx) => {
  return parseCircular(parse_helper_object, value, ctx);
};
const parse_helper_object = (value, ctx) => {
  const object = value;
  if (typeof object === "object" && object != null) {
    const newEntries = []; let value, decoded;
    if (hasProp(object, PROTO_KEY)) {
      ctx.addError("Object has own property __proto__");
      if (ctx.shouldAbort()) return ctx.returnErrors();
    }
    const extraneousKeys = new Set(Object.keys(object));
    ctx.push("a");
    value = getProp(object, "a");
    decoded = parse_null(value, ctx);
    if (decoded.ok) {
      newEntries.push(["a", decoded.value]);
    }
    ctx.pop();
    if (ctx.shouldAbort()) return ctx.returnErrors();
    extraneousKeys.delete("a");
    ctx.push("b");
    value = getProp(object, "b");
    decoded = parse_number(value, ctx);
    if (decoded.ok) {
      newEntries.push(["b", decoded.value]);
    }
    ctx.pop();
    if (ctx.shouldAbort()) return ctx.returnErrors();
    extraneousKeys.delete("b");
    ctx.push("c");
    value = getProp(object, "c");
    decoded = parse_array(value, ctx);
    if (decoded.ok) {
      newEntries.push(["c", decoded.value]);
    }
    ctx.pop();
    if (ctx.shouldAbort()) return ctx.returnErrors();
    extraneousKeys.delete("c");
    ctx.push("d");
    value = getProp(object, "d");
    decoded = parse_object0(value, ctx);
    if (decoded.ok) {
      newEntries.push(["d", decoded.value]);
    }
    ctx.pop();
    if (ctx.shouldAbort()) return ctx.returnErrors();
    extraneousKeys.delete("d");
    if (extraneousKeys.size > 0) {
      return reportExtraneousKeys(ctx, extraneousKeys);
    }
    return ctx.result(Object.fromEntries(newEntries));
  }
  return ctx.error("Value is not an object");
};
const parse_null = (value, ctx) => {
  return value === null ? ctx.result(value) : ctx.error("Value is not null");
};
const parse_number = (value, ctx) => {
  return typeof value === "number" ? ctx.result(value) : ctx.error("Value is not a number");
};
const parse_array = (value, ctx) => {
  return parseCircular(parse_helper_array, value, ctx);
};
const parse_helper_array = (value, ctx) => {
  if (Array.isArray(value)) {
    const newArray = [];
    for (let i = 0; i < value.length; i++) {
      ctx.push(i);
      const result = parse_string(value[i], ctx);
      if (result.ok) newArray.push(result.value);
      ctx.pop();
      if (ctx.shouldAbort()) break;
    }
    return ctx.result(newArray);
  }
  return ctx.error("Value is not an array");
};
const parse_string = (value, ctx) => {
  return typeof value === "string" ? ctx.result(value) : ctx.error("Value is not a string");
};
const parse_object0 = (value, ctx) => {
  return parseCircular(parse_helper_object0, value, ctx);
};
const parse_helper_object0 = (value, ctx) => {
  const object = value;
  if (typeof object === "object" && object != null) {
    const newEntries = []; let value, decoded;
    if (hasProp(object, PROTO_KEY)) {
      ctx.addError("Object has own property __proto__");
      if (ctx.shouldAbort()) return ctx.returnErrors();
    }
    const extraneousKeys = new Set(Object.keys(object));
    catchUnknownKeys(object, ctx, newEntries, extraneousKeys, parse_string, parse_bigint, false);
    return ctx.result(Object.fromEntries(newEntries));
  }
  return ctx.error("Value is not an object");
};
const parse_bigint = (value, ctx) => {
  return typeof value === "bigint" ? ctx.result(value) : ctx.error("Value is not a bigint");
};
",
  "entryFunc": "parse_object",
  "makeFunc": [Function],
}
`;

exports[`compile parse example 2`] = `
{
  "errors": [
    SchemaError {
      "message": "Value is not null",
      "path": Path {
        "context": null,
        "key": "a",
        "parent": Path {
          "context": null,
          "key": null,
          "parent": null,
        },
      },
    },
  ],
  "ok": false,
}
`;
